model-optimizer/extensions/ops/einsum.py

```diff
@@ -0,0 +1,248 @@
+    str, list, bool):
+        """
+        In order to satisfy NCHW layout, subscripts for tensors with rank greater 3 must be adjusted by moving labels
+        of the last dimension to the second position in the subscript. There is an exception for such tensors when
+        the label is ellipsis and it covers multiple tail dimensions. The method returns equation with adjusted subscripts
+        to NCHW layout along with a boolean mask to indicate which subscripts are adjusted.
+
+        :param node_name: Einsum node name for which equation is adjusted
+        :param equation: Equation to be adjusted
+        :param input_ranks: a list of input ranks
+        :param output_rank: output rank
+        :return: adjusted equation, boolean mask for inputs, and boolean flag if output subscript is adjusted
+        """
+        is_inputs_permuted = []
+        input_subscripts, output_subscript = Einsum.parse_equation(node_name, equation)
+        num_inputs = len(input_ranks)
+        assert len(input_subscripts) == num_inputs, "The number of inputs must match a number " \
+                                                    "of input subscripts"
+
+        # permute labels in input subscripts and mark inputs for which inference in NCHW layout is acceptable
+        # in case ellipsis covering multiple dimensions in the end, the permutation is impossible
+        # so the corresponding input must be in the original format (NHWC)
+        permuted_input_subscripts = []
+        for input_ind in range(num_inputs):
+            input_subscript = input_subscripts[input_ind]
+            input_rank = input_ranks[input_ind]
+            labels = Einsum.extract_subscript_labels(node_name, input_subscript)
+            num_broadcasted_dims = input_rank - len(labels) + 1
+            if input_rank > 3 and labels[-1] != "...":
+                is_inputs_permuted.append(True)
+                labels.insert(1, labels[-1])
+                del labels[-1]
+            elif input_rank > 3 and labels[-1] == "..." and num_broadcasted_dims == 1:
+                is_inputs_permuted.append(True)
+                labels.insert(1, labels[-1])
+                del labels[-1]
+            else:
+                is_inputs_permuted.append(False)
+            permuted_input_subscript = ''.join(labels)
+            permuted_input_subscripts.append(permuted_input_subscript)
+
+        # perform the same procedure for the output subscript as for the inputs subscripts
+        labels = Einsum.extract_subscript_labels(node_name, output_subscript)
+        num_broadcated_dims = output_rank - len(labels) + 1
+        if output_rank > 3 and labels[-1] != "...":
+            is_output_permuted = True
+            labels.insert(1, labels[-1])
+            del labels[-1]
+        elif output_rank > 3 and labels[-1] == "..." and num_broadcated_dims == 1:
+            is_output_permuted = True
+            labels.insert(1, labels[-1])
+            del labels[-1]
```

Comment:
may be fuse theese two branch in one ? for example if output_rank > 3 and ( labels[-1] != "..." or labels[-1] == "..." and num_broadcated_dims == 1)

Reply:
Done.